sdks/java/core/src/test/java/org/apache/beam/sdk/util/UnboundedScheduledExecutorServiceTest.java

```diff
@@ -502,4 +509,45 @@ void wakeUpAndCheckTasks(UnboundedScheduledExecutorService executorService) thro
     }
     Thread.sleep(100);
   }
+
+  @Test
+  public void testThreadsAreAddedOnlyAsNeededWithContention() throws Exception {
+    UnboundedScheduledExecutorService executorService = new UnboundedScheduledExecutorService();
+    CountDownLatch done = new CountDownLatch(1);
+
+    ThreadPoolExecutor executor =
+        new ThreadPoolExecutor(100, 100, Long.MAX_VALUE, MILLISECONDS, new SynchronousQueue<>());
+    // Schedule 1000 threads that are going to be scheduling work non-stop but sequentially.
+    for (int i = 0; i < 100; ++i) {
+      executor.execute(
+          () -> {
+            // Periodically check if done.
+            while (done.getCount() == 1) {
+              for (int j = 0; j < 100; ++j) {
+                try {
+                  executorService.submit(() -> {
+                    try {
+                      Thread.sleep(1);
+                    } catch (InterruptedException e) {
+                      throw new RuntimeException(e);
+                    }
+                  }).get();
+                } catch (InterruptedException | ExecutionException e) {
+                  // Ignore, happens on executor shutdown.
+                }
+              }
+            }
+          });
+    }
+
+    Thread.sleep(20 * 1000);
```

Comment:
Instead of using a sleep, consider having each thread wait for a condition so we know that they are all running and then each submit 1000 tasks.

Reply:
Done

Reply:
Much faster now :)
Passes with 99 threads with this change.
Failed with 108 without this change.

Reply:
I ran the test locally and it is flaking a lot (> 50%).

Without your change I see something like:
```
Feb 03, 2023 9:52:47 AM org.apache.beam.sdk.util.UnboundedScheduledExecutorServiceTest testThreadsAreAddedOnlyAsNeededWithContention
INFO: Created 105 threads to execute at most 100 parallel tasks
Feb 03, 2023 9:52:48 AM org.apache.beam.sdk.util.UnboundedScheduledExecutorServiceTest testThreadsAreAddedOnlyAsNeededWithContention
INFO: Created 107 threads to execute at most 100 parallel tasks
Feb 03, 2023 9:52:49 AM org.apache.beam.sdk.util.UnboundedScheduledExecutorServiceTest testThreadsAreAddedOnlyAsNeededWithContention
INFO: Created 105 threads to execute at most 100 parallel tasks
Feb 03, 2023 9:52:50 AM org.apache.beam.sdk.util.UnboundedScheduledExecutorServiceTest testThreadsAreAddedOnlyAsNeededWithContention
INFO: Created 105 threads to execute at most 100 parallel tasks
Feb 03, 2023 9:52:52 AM org.apache.beam.sdk.util.UnboundedScheduledExecutorServiceTest testThreadsAreAddedOnlyAsNeededWithContention
INFO: Created 103 threads to execute at most 100 parallel tasks
```

With your change I see:
```
Feb 03, 2023 9:54:20 AM org.apache.beam.sdk.util.UnboundedScheduledExecutorServiceTest testThreadsAreAddedOnlyAsNeededWithContention
INFO: Created 101 threads to execute at most 100 parallel tasks
Feb 03, 2023 9:54:21 AM org.apache.beam.sdk.util.UnboundedScheduledExecutorServiceTest testThreadsAreAddedOnlyAsNeededWithContention
INFO: Created 100 threads to execute at most 100 parallel tasks
Feb 03, 2023 9:54:22 AM org.apache.beam.sdk.util.UnboundedScheduledExecutorServiceTest testThreadsAreAddedOnlyAsNeededWithContention
INFO: Created 97 threads to execute at most 100 parallel tasks
Feb 03, 2023 9:54:24 AM org.apache.beam.sdk.util.UnboundedScheduledExecutorServiceTest testThreadsAreAddedOnlyAsNeededWithContention
INFO: Created 101 threads to execute at most 100 parallel tasks
Feb 03, 2023 9:54:25 AM org.apache.beam.sdk.util.UnboundedScheduledExecutorServiceTest testThreadsAreAddedOnlyAsNeededWithContention
INFO: Created 101 threads to execute at most 100 parallel tasks
```
and it flaked 29 times out of 50.

Reply:
Out of 100 I only saw at most 101 with my change, I modified the test to check <=104 to prevent flakes since that failed with previous. And we don't really need to be too strict here, just know it's better.